### PR TITLE
chore: improve diagnostics for invalid for loop annotation

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Improve diagnostics for invalid loop variable annotation and add tests for #3596.

Compiling this contract throws an error that does not make sense
```
@external
def foo():
    for i: uint9 in [1, 2, 3]:
        pass
```
```
vyper.exceptions.UnknownType: No builtin or user-defined type named '@exte'. 
  contract "tmp/old_for.vy:1", function "foo", line 1:0 
  ---> 1 @external
  -------^
       2 def foo():
```

This PR changes the error message to:
```
vyper.exceptions.UnknownType: No builtin or user-defined type named 'uint9'. Did you mean 'uint96', or maybe 'uint8'?
  contract "tmp/old_for.vy:3", function "foo", line 3:4 
       2 def foo():
  ---> 3     for i: uint9 in [1, 2, 3]:
  -----------^
       4         pass
```

### How I did it

Propagate the annotation source code and `For` node locations to the annotation and all its child nodes

### How to verify it

See tests.

### Commit message

```
chore: improve diagnostics for invalid for loop annotation
```

### Description for the changelog

Improve diagnostics for invalid for loop annotation

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/ast/parse.py [hunk 1]
```diff
@@ -258,6 +258,17 @@ def visit_For(self, node):
...
+        # this step improves the diagnostics during semantics analysis as otherwise
+        # the code displayed in the error message would be incorrectly based on the
+        # full source code with the location of the type annotation as an expression
+        annotation_children = python_ast.iter_child_nodes(node.target.annotation)
+        for n in [*annotation_children, node.target.annotation]:
+            # override the source code to show the spliced type annotation
+            n.node_source_code = raw_annotation
+
+            # override the locations to show the `For` node
+            python_ast.copy_location(n, node)
```
Comment (charles-cooper) [inline vyper/ast/parse.py:270]: i think `python_ast.increment_lineno` might be better here

**Review by charles-cooper — CHANGES_REQUESTED** [review]
nice new tests.

i think the annotation is a bit tricky to get right, since you're copying the location of the parent `for` node into the for target. here's a (contrived) example of where that doesn't work so well:
```vyper
@external
def foo():
    for i: \
        \
        \
        \
        uint9 in [1, 2, 3]:
        pass
```
with the current implementation that results in an error message something like
```vyper
vyper.exceptions.UnknownType: No builtin or user-defined type named 'uint9'. Did you mean 'uint96', or maybe 'uint8'?
  contract "repros/3721.vy:3", function "foo", line 3:4 
       2 def foo():
  ---> 3     for i: \
  -----------^
       4         \
```
i think what we really want to do is go into the tokens generated by ASTTokens (https://asttokens.readthedocs.io/en/latest/api-index.html#asttokens.ASTTokens.tokens) and modify all of them to have the correct start/end offsets.

File: vyper/ast/parse.py [hunk 2]
```diff
@@ -238,25 +241,50 @@ def visit_For(self, node):
...
+            return x[0] + y[0], x[1] + y[1]
+
+        for n in python_ast.walk(annotation):
+            # adjust all offsets
+            if hasattr(n, "first_token"):
+                n.first_token = n.first_token._replace(
+                    start=_add_pair(n.first_token.start, adjustment),
+                    end=_add_pair(n.first_token.end, adjustment),
+                    startpos=n.first_token.startpos + og_target.first_token.startpos,
+                    endpos=n.first_token.startpos + og_target.first_token.startpos,
```
Comment (tserg) [inline vyper/ast/parse.py:276]:
```suggestion
                    endpos=n.first_token.endpos + og_target.first_token.startpos,
```
Should this be `endpos`?

Comment (tserg) [reply vyper/ast/parse.py:276]: False alarm.

**Comment by charles-cooper** [comment]
i have a solution for https://github.com/vyperlang/vyper/pull/3721#pullrequestreview-1809505443. i'll submit it in a follow up PR.

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**